### PR TITLE
feat:  Adds support for -server mode and concurrently downloading dashboards

### PR DESCRIPTION
**Pull request by vanugrah**
This is a useful script. However it is a pain to manually re-run the script to keep track of a migration. This PR: 

- Adds an HTTP Server which returns the list of angular dashboard as a JSON response on the `/output` route.  
- Adds a flag to configure the server port (default 8080). 
- Setups up a background go routine to update on a specified interval (default 10m). 
- Adds a flag to specify custom refresh interval. 
- Moves flag parsing into a new package. 
- Extends LeveledLogger with an ErrorLogger. 
- Adds readiness probe since detections can take 5+ minutes to run. 
- Adds signal handling. 
- Adds support for concurrently download dashboards to speed up detection runs (default 5) 
- General tidying.

**This can now be hooked up directly to Grafana via the infinity data source plugin for live migration tracking.**

Usage: 
```
detect-angular-dashboards -server -interval 5m -max-concurrency 10 https://<grafana_instance>/api
```

Note: There was a strange bug with using  JSONOutputter.Output to filter the angular dashboards which resulted in a increasing number of duplicates with each request to the /output endpoint. My suspicion is its something to do with in place filtering of the Dashboards slice. I eventually decided to write my own version which seems to be working as expected `filterAngularDashboards`.

### CODE REVIEW

**Review by xnyo — COMMENTED** [review]
Thanks for your contribution! Awesome work on this!! :raised_hands: :rocket: 

I really like the idea of having an HTTP server and displaying the output within Grafana with the Infinity plugin.

The code looks good overall, but I have left some comments and suggestions which should be addressed before we can merge

**Comment by xnyo** [comment]
I will also try to debug the issue with  JSONOutputter.Output to filter the angular dashboards that's causing some duplicates after each request :slightly_smiling_face:

File: main.go
Comment (vanugrah) [reply main.go:84]: Thank you for this! This was on my list to address, but I got distracted with other work. The PR looks good - but since we've gone back to using the `detector.Run` function, I fear this may re-introduce the JSONOutputter.Output duplicate bug. Happy to merge your PR into this branch once we have that addressed.

File: main.go
Comment (vanugrah) [reply main.go:93]: This is mainly to play nicely when deployed in kubernetes. I like the pattern of having a separate `/ready` endpoint as it allows flexibility for more nuanced readiness logic.  The main constraint of the readiness probe is that it has to be available on startup and explicitly fail until the first detection has finished and we have results that we can return.

File: main.go
Comment (vanugrah) [reply main.go]: Sorry I didn't  read your PR correctly. It worked as expected and I was able to merge it.

**Comment by vanugrah** [comment]
Hey @xnyo  - thanks a lot for the review. I have incorporated the majority of your feedback and merged in your PR with the changes to still enable the CLI mode. 

Only net new changes are a `runServer` method which sets up the HTTP server with signal handling. As far as I can tell http.ListenAndServe actually does not setup signal handlers appropriately.  

Can you please take another look?

File: flags/flags.go
Comment (xnyo) [inline flags/flags.go:27]: I think this needs to explain that the value must be a listen address, and remove the interval reference as that's a separate argument anyways:

```suggestion
	flag.StringVar(&flags.Server, "server", "", "Run as HTTP server instead of CLI. Value must be a listen address (e.g.: 0.0.0.0:5000. Output is exposed as JSON at /output.")
```

File: flags/flags.go
Comment (xnyo) [inline flags/flags.go:26]: Nit:

```suggestion
	flag.DurationVar(&flags.Interval, "interval", 5*time.Minute, "detection refresh interval when running in HTTP server mode")
```

File: main.go
Comment (xnyo) [inline main.go:112]: Nit: After taking another look at this, I think it would make more sense to rename this handler to `/detections`. WDYT?

```suggestion
	http.HandleFunc("/detections", func(w http.ResponseWriter, r *http.Request) {
		handleDetectionsRequest(w, r, &out, log)
	})
```

**Review by xnyo — COMMENTED** [review]
Thanks for addressing the comments, awesome work! :raised_hands: :rocket:  Just a couple of very small things I have noticed mostly around naming:

**Comment by vanugrah** [comment]
Hey @xnyo  - Agree with all of your suggestions. I've updated the PR accordingly.

File: flags/flags.go
Comment (xnyo) [inline flags/flags.go:27]: Only one thing left:

```suggestion
	flag.StringVar(&flags.Server, "server", "", "Run as HTTP server instead of CLI. Value must be a listen address (e.g.: 0.0.0.0:5000. Output is exposed as JSON at /detections.")
```

**Comment by vanugrah** [comment]
Done - sorry I have not been looking at this PR very closely. Thanks for your thorough review.

**Review by xnyo — APPROVED** [review]
LGTM, awesome work! :rocket:  Thank you for contributing!

**Pull request merged**
